File: leetcode/Python/0039-m-combination-sum.py
"""
DFS
"""
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        self.dfs(sorted(candidates), [], target, 0, res)
        return res

    def dfs(self, candidates, sublist, target, start, res):
        if target == 0:
            res.append(sublist[:])
        for i in range(start, len(candidates)):
            if candidates[i] > target:
                return
            sublist.append(candidates[i])
            self.dfs(candidates, sublist, target - candidates[i], i, res)
            sublist.pop()

File: leetcode/C++/0039-m-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> path;
        dfs(candidates, target, 0, path, res);
        return res;
    }
    
    void dfs(vector<int>& candidates, int target, int start, 
             vector<int>& path, vector<vector<int>>& res) {
        if (target == 0)
            res.push_back(path);
        for (int i = start; i < candidates.size(); i++) {
            if (candidates[i] > target)
                return;
            path.push_back(candidates[i]);
            dfs(candidates, target - candidates[i], i, path, res);
            path.pop_back();
        }
    }
};